app/models/post.rb

```diff
@@ -41,6 +41,7 @@ class Post < ActiveRecord::Base
   has_many :post_actions, dependent: :destroy
   has_many :topic_links
   has_many :group_mentions, dependent: :destroy
+  belongs_to :bookmarkable_relation, class_name: "Topic", foreign_key: :topic_id
```

Comment:
I have to explain this because it's a little weird...in plugins like assign and encrypt, we have to load additional data for the topic. For the Topic bookmarks we can do `Bookmark.select_type(bookmarks, 'Topic').map(&:bookmarkable)` to get the topics but for the post ones we have to do something like this `Bookmark.select_type(bookmarks, 'Post').map { |bm| bm.bookmarkable.topic }`. To make the latter more generic, I think we can allow things that will be attached to bookmarks one additional relation to make things like this easier `Bookmark.select_type(bookmarks, 'Post').map { |bm| bm.bookmarkable.bookmarkable_relation }`.

The relation and the bookmarkable are preloaded here:

https://github.com/discourse/discourse/pull/16335/files#diff-0f7c6a210ef52b10b632c103844a35108a837501df63a47d2f7e077cf95b878fR106

Reply:
Does this mean any model that is bookmarkable will have to implement this relation?

Reply:
No not any model...it's kind of tricky, the only case where it happens right now is because we need to get the bookmarkable (post)'s topic record for certain plugin things...I just wanted to make it more generic so we're not explicitly calling bookmarkable.topic which seems a little weird.

Reply:
@tgxworld also did you see this comment?

Reply:
Yup I saw this comment but the bookmarks on post are already somewhat special and handled as such. I'm not sure if we really need `bookmarkable_relation` here instead of just using `topic`.